service/geopmdpy/varrun.py

```diff
@@ -700,11 +661,98 @@ def _load_session_file(self, sess_path):
             if batch_pid is not None and \
                self._is_pid_valid(batch_pid, file_time) == False:
                 sess.pop('batch_server')
+            self._sessions[client_pid] = dict(sess)
+            self._update_session_file(client_pid)
         else:
-            # Invalid session, remove batch server
-            sess['client_pid'] = self._INVALID_PID
-            if batch_pid is not None:
-                sess.pop('batch_server')
+           os.rename(sess_path, renamed_path)
 
-        self._sessions[client_pid] = dict(sess)
-        self._update_session_file(client_pid)
+
+class WriteLock(object):
+    def __init__(self, var_path=GEOPM_SERVICE_VAR_PATH):
+        self._VAR_PATH = var_path
+        self._LOCK_PATH = os.path.join(self._VAR_PATH, "CONTROL_LOCK")
+        self._fid = None
+
+    def __enter__(self):
+        old_mask = os.umask(0o077)
+        try:
+            trial_count = 0
+            while self._fid is None and trial_count < 2:
+                if os.path.exists(self._LOCK_PATH):
+                    # Rename exiting file if it is not secure
+                    is_secure_path(self._LOCK_PATH)
+                self._fid = open(self._LOCK_PATH, 'a+')
+                if not is_secure_file(self._LOCK_PATH, self._fid):
+                    # File is insecure and was renamed, try again
+                    self._fid.close()
+                    self._fid = None
+                trial_count += 1
+        finally:
+            os.umask(old_mask)
+        if self._fid == None:
+            raise RuntimeError('Unable to open write lock securely after two tries')
+        fcntl.lockf(self._fid, fcntl.LOCK_EX)
+        return self
+
+    def __exit__(self, type, value, traceback):
+        if self._fid is not None:
+            fcntl.lockf(self._fid, fcntl.LOCK_UN)
+            self._fid.close()
+
+    def try_lock(self, pid=None):
+        """Get the PID that holds the lock or set lock
+
+        Returns the PID that currently holds the lock.  If the user specifies
+        a PID and the lock is not held by another process, then the lock will
+        be assigned, and the input pid will be returned.  The user must check
+        the return value to determine if a request to assign the lock was
+        successful, an error is not raised if the lock is held by another PID.
+
+        None is returned if the write lock is not held by any active session
+        and the user does not specify a pid.
+
+        Args:
+            pid (int): The PID to assign the lock
+
+        Returns:
+            int: The PID of the session that holds the write lock upon return
+                 or None if the write lock is not held
+
+        """
+        file_pid = None
+        self._fid.seek(0, os.SEEK_END)
```

Comment:
Why seek() instead of readline()?  It might simplify this logic.

Reply:
Updated implementation, should be a bit easier to read.  Thanks